### DISCUSSION TITLE
Great update

### DISCUSSION THREAD

**Discussion by Zeyadas** [koreader/koreader#7187] [General]
The last update 1.2021 is great guys..
The new icons is beautiful and I noticed the (faster) responses after using it for 5 minutes. 
So many thanks..
EDIT: I'm using an android tablet.

**Comment by NiLuJe** [comment]
Hopefully I didn't break it in the dreaded Android 4.0 - 4.x range, but I think we've already seen success reports on Tolinos ;).

**Reply by Zeyadas** [reply]
Mine is 5.1.1
And it's much faster than the previous versions. 
Can't wait for the next version 😉.

**Comment by elvvis** [comment]
I can only confirm that, fantastic work! It runs so smoothly on the Tolino now, no comparison to before when it sometimes took up to 10 seconds for an action.

**Comment by NiLuJe** [comment]
On another Android note, there was a single report that enabling night mode would crash, something that made no sense whatsoever to me on a technical level, so, if anyone can reproduce that, that'd be an interesting data point ;).

(xref and workaround: https://www.mobileread.com/forums/showpost.php?p=4077266&postcount=335)

**Reply by elvvis** [reply]
I can confirm that it crashes. With the option „Disable C Blitter“ it works. Is there already an issue or should I create a new one?

**Reply by NiLuJe** [reply]
No issue yet, and, ideally, I'd like someone to reproduce it on another device than a crappy Tolino an Android 4 before that ;).

**Reply by NiLuJe** [reply]
I might throw you a debug build with more logging at one point if that doesn't happen though, but, until it gets reproduced in a saner environment, it's fairly low priority right now.

**Reply by NiLuJe** [reply]
Note to self: The only vague thing that comes to mind would be that the C version somehow ends up generating NEON code, although that shouldn't really be happening for those Android builds.

@elvvis: Which device is it, exactly, and do you have any details about the SoC/CPU it's using?

@pazos: Is there an easy way to discriminate a SIGSEGV from a SIGILL on Android? Because those 'app has died' logs are thoroughly unhelpful ^^.

**Reply by Frenzie** [reply]
Works fine on all of my devices plus emulators afaict.

**Reply by Frenzie** [reply]
@NiLuJe are you using the debug build? Otoh it just says that

**Reply by NiLuJe** [reply]
> @NiLuJe are you using the debug build? Otoh it just says that

Says what? The SoC's name?

**Reply by Frenzie** [reply]
No, the error code sigsegv vs sigill stuff.

**Reply by Frenzie** [reply]
Wrt soc Android has always had a device info screen.

![Screenshot_2021-01-23-15-48-24-63](https://user-images.githubusercontent.com/202757/105591390-9e74df80-5d92-11eb-8b57-d6a3ce8f1a24.jpg)

**Reply by elvvis** [reply]
> @elvvis: Which device is it, exactly, and do you have any details about the SoC/CPU it's using?

I am using a Tolino Epos 2, Android 4.4.2.

**Reply by NiLuJe** [reply]
Okay, so, essentially a Kobo Forma, but with a crappier OS ;p.

Which kinda throws my code generation theory out the window.

I'll try to get you a debug build with some more logging later today ;).

**Reply by NiLuJe** [reply]
Hmm, lightbulb moment. If the Android widow buffer is RGB565, this will indeed break. I'll get a debug build ready to confirm ;).

**Reply by NiLuJe** [reply]
Okay, [here comes a debug build](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug.zip) that should be more chatty ;).

You can install it in parallel with your current release build, it'll have an orange icon (which'll hopefully translate in a shade lighter/darker on eInk ^^).

Once that's done, check the logs via `adb logcat 'KOReader:* ActivityManager:* AndroidRuntime:* DEBUG:* *:F dlopen:* LuaJIT:* KOCBB:*'`.

Ideally, I'll need the full log from a session with the C BB enabled (e.g., restart, and just toggle nightmode).

**Reply by NiLuJe** [reply]
For posterity's sake, the actual patch to base:

```diff
diff --git a/Makefile b/Makefile
index 4b66b4b0..204710cd 100644
--- a/Makefile
+++ b/Makefile
@@ -205,7 +205,7 @@ ifdef DARWIN
 endif
 
 $(OUTPUT_DIR)/libs/libblitbuffer.so: blitbuffer.c
-       $(CC) $(DYNLIB_CFLAGS) $(VECTO_CFLAGS) $(SYMVIS_FLAGS) -o $@ $^
+       $(CC) $(DYNLIB_CFLAGS) $(VECTO_CFLAGS) $(SYMVIS_FLAGS) -o $@ $^ -landroid -llog
 
 $(OUTPUT_DIR)/libs/libwrap-mupdf.so: wrap-mupdf.c \
                        $(MUPDF_LIB)
diff --git a/blitbuffer.c b/blitbuffer.c
index c063cf1f..adb19bd0 100644
--- a/blitbuffer.c
+++ b/blitbuffer.c
@@ -22,6 +22,9 @@
 #include <string.h>
 #include "blitbuffer.h"
 
+#include <android/log.h>
+#define CBB_LOG(fmt, ...) ((void)__android_log_print(ANDROID_LOG_VERBOSE, "KOCBB", "%s: " fmt, __FUNCTION__, ##__VA_ARGS__))
+
 static const char*
     get_bbtype_name(int bbtype)
 {
@@ -2136,9 +2139,13 @@ void BB_invert_blit_from(BlitBuffer * restrict dst, const BlitBuffer * restrict
     const int sbb_type = GET_BB_TYPE(src);
     const int sbb_rotation = GET_BB_ROTATION(src);
     const int dbb_rotation = GET_BB_ROTATION(dst);
+    CBB_LOG("dst: %p, src: %p, dest_x: %u, dest_y: %u, offs_x: %u, offs_y: %u, w: %u, h: %u",
+                dst, src, dest_x, dest_y, offs_x, offs_y, w, h);
+    CBB_LOG("dbb_type: %s, sbb_type: %s, sbb_rotation: %d, dbb_rotation: %d",
+                get_bbtype_name(dbb_type), get_bbtype_name(sbb_type), sbb_rotation, dbb_rotation);
     if (dbb_type != sbb_type) {
-        fprintf(stderr, "%s: incompatible bb (dst: %s, src: %s) in file %s, line %d!\n",
-                __FUNCTION__, get_bbtype_name(dbb_type), get_bbtype_name(sbb_type), __FILE__, __LINE__);
+        CBB_LOG("incompatible bb (dst: %s, src: %s) in file %s, line %d!",
+                get_bbtype_name(dbb_type), get_bbtype_name(sbb_type), __FILE__, __LINE__);
         exit(1);
     }
     switch (dbb_type) {
diff --git a/ffi/framebuffer_android.lua b/ffi/framebuffer_android.lua
index 6d805e3d..8053efe5 100644
--- a/ffi/framebuffer_android.lua
+++ b/ffi/framebuffer_android.lua
@@ -47,6 +47,7 @@ end
 function framebuffer:init()
     -- we present this buffer to the outside
     self.bb = BB.new(android.screen.width, android.screen.height, BB.TYPE_BBRGB32)
+    android.LOGV(string.format("KOReader buffer is RGB32: %dx%d %s", android.screen.width, android.screen.height, tostring(self.bb.data)))
     self.bb:fill(BB.COLOR_WHITE)
     self:_updateWindow()
     framebuffer.parent.init(self)
@@ -117,8 +118,10 @@ function framebuffer:_updateWindow()
     or buffer[0].format == C.WINDOW_FORMAT_RGBX_8888
     then
         bb = BB.new(buffer[0].width, buffer[0].height, BB.TYPE_BBRGB32, buffer[0].bits, buffer[0].stride*4, buffer[0].stride)
+        android.LOGV(string.format("Window buffer is RGB32: %dx%d (stride: %d, pixel_stride: %d)", tonumber(buffer[0].width), tonumber(buffer[0].height), tonumber(buffer[0].stride*4), tonumber(buffer[0].stride)))
     elseif buffer[0].format == C.WINDOW_FORMAT_RGB_565 then
         bb = BB.new(buffer[0].width, buffer[0].height, BB.TYPE_BBRGB16, buffer[0].bits, buffer[0].stride*2, buffer[0].stride)
+        android.LOGV(string.format("Window buffer is RGB565: %dx%d (stride: %d, pixel_stride: %d)", tonumber(buffer[0].width), tonumber(buffer[0].height), tonumber(buffer[0].stride*2), tonumber(buffer[0].stride)))
     else
         android.LOGE("unsupported window format!")
     end
@@ -136,6 +139,7 @@ function framebuffer:_updateWindow()
             -- so, simply invert everything *now* ;).
             -- The idea is that we absolutely want to avoid the Lua BB on Android, because it is *extremely* erratic,
             -- because of the mcode alloc issues...
+            android.LOGV(string.format("NightMode: invertblitFrom %s to %s", tostring(ext_bb.data), tostring(bb.data)))
             bb:invertblitFrom(ext_bb)
         else
             bb:blitFrom(ext_bb)
```

**Reply by pazos** [reply]
Hmm, maybe [taking window ownership](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/MainActivity.kt#L96-L105) changes the buffer format?

That could explain why it fails on tolino.

In that case using `window.setFormat(PixelFormat.RGBA_8888)` just before `setContentView(view)` should change the buffer to default format used elsewhere.

If you want to test that yourself you can add edit https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt#L271 to `NEEDS_VIEW = EINK_FREESCALE || EINK_QCOM || EMULATOR_X86` and test on a emulator between 17 and 28 :+1

**Reply by NiLuJe** [reply]
I don't know if we can force the HW buffer to a different format (which would be preferable here, if only for performance reasons because RGB565 is the worst), but I suspect that it's currently RGB565 on Tolinos, while our own buffer is always RGB32, and the CBB's invertBlitFrom just doesn't support cross-bitdepth scenarios, and just throws an error and exit(1) ;).

**Reply by NiLuJe** [reply]
@pazos: I don't have any emulators (and I don't really want some ;p), so, how would I go about testing that either on a real device on API 29, or a real device on whatever-API-Android -7-is ^^.

**Reply by NiLuJe** [reply]
@elvvis: Alternatively, [this build](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug-r1.zip) kept the same fancy logging as above, but should *also* workaround the issue.

(i.e., it includes https://github.com/koreader/koreader-base/pull/1293)

**Reply by elvvis** [reply]
> 
> 
> Okay, [here comes a debug build](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug.zip) that should be more chatty ;).
> 
> You can install it in parallel with your current release build, it'll have an orange icon (which'll hopefully translate in a shade lighter/darker on eInk ^^).
> 
> Once that's done, check the logs via `adb logcat 'KOReader:* ActivityManager:* AndroidRuntime:* DEBUG:* *:F dlopen:* LuaJIT:* KOCBB:*'`.
> 
> Ideally, I'll need the full log from a session with the C BB enabled (e.g., restart, and just toggle nightmode).

Here is a log with C BB enabled:
[log.txt](https://github.com/koreader/koreader/files/5860935/log.txt)

**Reply by NiLuJe** [reply]
@elvvis: Yay! Hunch confirmed ;). Thanks!

```
V/KOCBB   ( 3289): BB_invert_blit_from: incompatible bb (dst: BBRGB16, src: BBRGB32) in file blitbuffer.c, line 2148!
```

The second build I posted right before your answer should "work" (caveats explained in the PR). You'll have to uninstall the previous debug build first, though.

**Reply by elvvis** [reply]
> 
> 
> @elvvis: Alternatively, [this build](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug-r1.zip) kept the same fancy logging as above, but should _also_ workaround the issue.
> 
> (i.e., it includes [koreader/koreader-base#1293](https://github.com/koreader/koreader-base/pull/1293))

Yes, this version works fine

**Reply by pazos** [reply]
> @pazos: I don't have any emulators (and I don't really want some ;p), so, how would I go about testing that either on a real device on API 29, or a real device on whatever-API-Android -7-is ^^.

Sure. Just use https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/MainActivity.kt#L97-L101 if you want your device to behave like a tolino.

**Reply by pazos** [reply]
Or just ignore my comments if your PR does the job ;)

**Reply by NiLuJe** [reply]
I mean, this might be tricky to test because I'd admittedly be doing something stupid (switching *from* RGB32 to RGB565 ;p), but, eh, worth a try ;).

Would it work on API 29 or do I really need something older?

**Reply by NiLuJe** [reply]
@elvvis: Yay! Stupid question: when in nightmode, are the "colors" visibly off, or do you really need to know that they're off to notice it? ^^.

**Reply by NiLuJe** [reply]
@pazos: Why, yes, yes, it works ;p.

```
Window buffer is RGB565: 1080x2200 (stride: 2176, pixel_stride: 1088)
```

**Reply by NiLuJe** [reply]
Which means...

@elvvis: Does [this build](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug-r2.zip) work (at all ^^), and, whether it does or not, could you attach the same kind of logs as before?

**Reply by NiLuJe** [reply]
(Diff for posterity)

```diff
diff --git a/app/src/main/java/org/koreader/launcher/MainActivity.kt b/app/src/main/java/org/koreader/launcher/MainActivity.kt
index 768f2ed..36bb622 100644
--- a/app/src/main/java/org/koreader/launcher/MainActivity.kt
+++ b/app/src/main/java/org/koreader/launcher/MainActivity.kt
@@ -8,6 +8,7 @@ import android.content.Context
 import android.content.Intent
 import android.content.IntentFilter
 import android.content.pm.PackageManager
+import android.graphics.PixelFormat
 import android.net.Uri
 import android.os.BatteryManager
 import android.os.Build
@@ -97,6 +98,9 @@ class MainActivity : NativeActivity(), JNILuaInterface,
             view = NativeSurfaceView(this)
             window.takeSurface(null)
             view?.holder?.addCallback(this)
+            // Convince Android to use RGB32, because some devices still prefer RGB565 surfaces,
+            // which is generally slower since we do all our blitting on the CPU.
+            window.setFormat(PixelFormat.RGBA_8888)
             setContentView(view)
             takesWindowOwnership = true
             "SurfaceView"
diff --git a/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt b/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt
index 56cfdcb..1870370 100644
--- a/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt
+++ b/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt
@@ -268,7 +268,8 @@ object DeviceInfo {
         BUG_SCREEN_ROTATION = BUG == BugDevice.EMULATOR
 
         // needs a surfaceView to do epd updates
-        NEEDS_VIEW = EINK_FREESCALE || EINK_QCOM
+        // We also enable it on *all* Tolinos to be able to enforce RGB32 HW buffers.
+        NEEDS_VIEW = EINK_FREESCALE || EINK_QCOM || TOLINO_EPOS
     }
 
     private fun getBuildField(fieldName: String): String {
diff --git a/jni/luajit/luajit b/jni/luajit/luajit
--- a/jni/luajit/luajit
+++ b/jni/luajit/luajit
@@ -1 +1 @@
-Subproject commit 1e66d0f9e6698fdee672c40a9a5b4159c9254091
+Subproject commit 1e66d0f9e6698fdee672c40a9a5b4159c9254091-dirty
```

**Reply by NiLuJe** [reply]
@pazos: Any reason why `TOLINO_EPOS` wasn't in the `EINK_FREESCALE` bunch already? (Or did I misunderstand that? ^^).

(I mean, @elvvis's device, the Epos 2, is essentially a Forma, which definitely uses an i.MX 6SLL ;)).

**Reply by Galunid** [reply]
@NiLuJe Unless I'm missing something, it's already included in `TOLINO`, and `TOLINO` is in `EINK_FREESCALE`

**Reply by pazos** [reply]
All tolinos are whitelisted. If not they don't work at all. About the APIs to test. My first hint was to use the EMULATOR_X86 and only these apis are recognized as such. If you're in a device and you enable the nativeSurfaceView trick manually then it should work in any supported API.

**Reply by NiLuJe** [reply]
> All tolinos are whitelisted.

Not sure if I follow ^^. In the specific context of the diff I posted above, does that mean than the added `TOLINO_EPOS` check is redundant? Because otherwise I'm not quite seeing how a `TOLINO_EPOS` would get flagged as `NEEDS_VIEW`?
(I'm assuming the brand/model strings are fairly unique, and the results of those checks only flag a RakutenKobo Tolino as TOLINO_EPOS and not TOLINO, which might be a completely bogus assumption, explaining my confusion ;)).

> My first hint was to use the EMULATOR_X86 and only these apis are recognized as such. If you're in a device and you enable the nativeSurfaceView trick manually then it should work in any supported API.

Yeah, I ended up just setting NEEDS_VIEW to true unilaterally to test it ;p.

**Reply by pazos** [reply]
[`TOLINO_EPOS`](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt#L185-L188) is used for the lights interface in . It should only match Tolino Epos1 and 2

[`TOLINO`](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt#L179-L183) is used for eink. It should match all tolinos with FW 11+

`TOLINO` is also [`EINK_FREESCALE`](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt#L234-L239), of the specific kind that need to drive [each single EPD update](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/device/DeviceInfo.kt#L262)

So basically what @Galunid said :) But your change doesn't hurt either. It just happens that *alone* doesn't make a tolino epos an eink device.

> Yeah, I ended up just setting NEEDS_VIEW to true unilaterally to test it ;p.

That works too ;)

**Reply by elvvis** [reply]
> 
> 
> Which means...
> 
> @elvvis: Does [this build](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug-r2.zip) work (at all ^^), and, whether it does or not, could you attach the same kind of logs as before?

This build also works. Here is the log for this build:
[log2.txt](https://github.com/koreader/koreader/files/5862074/log2.txt)

**Reply by NiLuJe** [reply]
@elvvis: That's a good start, but I'm not seeing any `KOReader` entries in your log (which is what would tell me if the RGB32 change took ;)).

Are you sure you used the exact logcat string I quoted [above](https://github.com/koreader/koreader/discussions/7187#discussioncomment-304950)?

**Reply by elvvis** [reply]
I used the same string as above and I also recorded the same actions as before: Start KOReader (launches into the file browser) and switch on night mode.

**Reply by NiLuJe** [reply]
Huh. I'd noticed it the first time around, actually, but the interesting bit wasn't tagged `KOReader` at the time, so it didn't matter ;p.

Do you see anything tagged like that if you run an unfiltered logcat? (might be hard to tell, because it's going to be *extremely* verbose).

Or, on the off-chance it's a weird CLI quirk, just `adb logcat KOReader:*`?

**Reply by NiLuJe** [reply]
Or even `adb logcat KOReader:V` to make very sure the shell isn't doing something stupid with that wildcard token ;).

**Reply by NiLuJe** [reply]
Or, even unlikelier, that you don't have debug logging & verbose debug logging disabled in Tools > More Tools > Developer Options.

(I don't *think* this should matter in a debug build, but, who knows ^^).

**Comment by elvvis** [comment]
After trying for the last hour to get a new log, I now give up. Something causes my system directory to be written full after starting KOReader (i used version "ko-android-2021.01-nm-rgb565-debug-r2"), causing KOReader to crash.
This may also be the reason why there is nothing useful in the last log.

**Reply by pazos** [reply]
Umm, logs are not written to disk. They're kept in a log buffer, so they can hardly fill your disk.

It could be that you have enough space for normal KOReader, but the debug version takes much more space (x2-x3). Even worse: most of the space is filled up on the first run, when assets are uncompressed.

For the purpose of testing this feature it should work with a release build, unless I miss something. Pinging @NiLuJe

**Reply by NiLuJe** [reply]
Well, those logs happen to be flagged verbose, but don't go through logger, so, yeah, that should work, lemme double-check ;).

**Reply by NiLuJe** [reply]
Yeah, that works, so [here comes r3](https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/ko-android-2021.01-nm-rgb565-debug-r3.zip) ;).

You will have to uninstall your existing version first, because this one is signed with a debug key.

(@pazos: That's ostensibly why I started with debug builds, FWIW).

**Comment by elvvis** [comment]
Here is a new log from r3 ("adb logcat KOReader:V").
[log3.txt](https://github.com/koreader/koreader/files/5863000/log3.txt)

**Reply by NiLuJe** [reply]
Hmm, well, that patch apparently didn't work out :'(.

```
V/KOReader( 2964): KOReader buffer is RGB32: 1440x1920 ColorRGB32(0, 0, 0, 0)
V/KOReader( 2964): Window buffer is RGB565: 1440x1920 (stride: 2880, pixel_stride: 1440)
```

Sidebar: Debug logging still seems to be enabled, is it checked in Tools > More tools > Developer options?

**Reply by elvvis** [reply]
> Sidebar: Debug logging still seems to be enabled, is it checked in Tools > More tools > Developer options?

Yes, debug logging is still enabled.

**Reply by NiLuJe** [reply]
Okay, so everything makes sense then, and that last patch just can't manage to convince those devices to use a sane pixel format :/.

Oh, well, at least the workaround behaves ;).

Many thanks for your tests and perseverance, that was very helpful :).

**Reply by pazos** [reply]
There are plenty of native APIs in https://developer.android.com/ndk/reference/group/a-native-window but, unless you're a true warrior of annoying remote-assisted debugging, I would stick with the workaround and forget about it :smile:

**Reply by NiLuJe** [reply]
Yep, definitely ;p.

Too bad, especially since it worked on my phone, but, oh, well ;).

**Comment by ichnilatis-gr** [comment]
Could the pencil mark in the pop-up dictionary be moved near the word entry, please? It's too hard to tap on it. Most of the time, when I try to tap on it, the pop-up window closes.

Thanks!

**Reply by poire-z** [reply]
> Could the pencil mark in the pop-up dictionary be moved near the word entry, please? 

I'd rather not. We put it there on the far right so it does not attract the eyes too much when reading the definition - and it's nice there with the other cosmetics stuff we don't care much about.

<kbd>![image](https://user-images.githubusercontent.com/24273478/106173138-e7160980-6193-11eb-836c-3901c8bd4b5c.png)</kbd>

Yes, it's near the close button. But I guess what happens to you is you hit outside the window, on the right of the icon, so closing it. May be see at hitting a bit on its left and try to get the habit...

**Comment by 1xCasino1az** [comment]
Thanks for the feedback! We're glad you're enjoying the update and noticing the improvements! 🙌

**Comment by 1xCasino1az** [comment]
Great to hear! It's awesome that the new icons and faster responses are making a difference for you.

**Comment by 1xCasino1az** [comment]
Thanks for sharing your device info! We'll keep working to improve the experience for Android users too!
